File: compiler/nodes/expressions/unit.go
package expressions

import (
	"wipple/codegen"
	"wipple/database"
	"wipple/syntax"
	"wipple/typecheck"
	"wipple/visit"
)

type UnitExpressionNode struct {
	Facts *database.Facts
}

func (node *UnitExpressionNode) GetFacts() *database.Facts {
	return node.Facts
}

func ParseUnitExpression(parser *syntax.Parser) (*UnitExpressionNode, *syntax.Error) {
	span := parser.Spanned()

	_, err := parser.Token("LeftParenthesis", syntax.TokenConfig{
		Reason: "between these parentheses",
	})
	if err != nil {
		return nil, err
	}

	_, err = parser.Token("RightParenthesis")
	if err != nil {
		return nil, err
	}

	return &UnitExpressionNode{
		Facts: database.NewFacts(span()),
	}, nil
}

func (node *UnitExpressionNode) Visit(visitor *visit.Visitor) {
	visitExpression(visitor, node)

	visitor.Constraint(typecheck.TypeConstraint(node, typecheck.TupleType([]database.Node{})))
}

func (node *UnitExpressionNode) Codegen(c *codegen.Codegen) error {
	span := database.GetSpanFact(node)
	c.WriteString(span, "[]")
	return nil
}

File: compiler/src/nodes/expressions/unit.rs
use crate::{
    codegen::{Codegen, CodegenCtx, CodegenResult, ir},
    database::{Node, NodeRef},
    nodes::visit_expression,
    syntax::{ParseError, Parser, TokenKind},
    typecheck::TypeConstraint,
    visit::{Visit, Visitor},
};

#[derive(Debug)]
pub struct UnitExpressionNode;

impl Node for UnitExpressionNode {}

pub fn parse_unit_expression(parser: &mut Parser<'_>) -> Result<UnitExpressionNode, ParseError> {
    parser.token_with_reason(TokenKind::LeftParenthesis, "between these parentheses")?;
    parser.token(TokenKind::RightParenthesis)?;

    Ok(UnitExpressionNode)
}

impl Visit for UnitExpressionNode {
    fn visit(&self, node: &NodeRef, visitor: &mut Visitor<'_>) {
        visit_expression(node, visitor);

        visitor.constraint(TypeConstraint::new(node.clone(), visitor.unit_type()));
    }
}

impl Codegen for UnitExpressionNode {
    fn codegen(&self, node: &NodeRef, ctx: &mut CodegenCtx<'_>) -> CodegenResult {
        ctx.instruction(ir::Instruction::Value {
            node: node.clone(),
            value: ir::Value::Tuple(Vec::new()),
        });

        Ok(())
    }
}
